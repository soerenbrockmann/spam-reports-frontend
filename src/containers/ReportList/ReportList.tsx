import React from 'react';
import { connect } from 'react-redux';
import {
    Report,
    Element,
    fetchReports,
    blockMessage,
    resolveMessage,
} from '../../actions';
import { StoreState } from '../../reducers';
import { ListItem } from '../../components/ListItem';

interface ReportListProps {
    reports: Report;
    fetchReports: Function;
    blockMessage: Function;
    resolveMessage: Function;
}

interface ReportListState {
    fetching: boolean;
}

class _ReportList extends React.Component<ReportListProps> {
    state: ReportListState = { fetching: true };

    componentDidMount() {
        this.props.fetchReports();
    }

    componentDidUpdate(prevProps: ReportListProps): void {
        if (
            !prevProps.reports.elements.length &&
            this.props.reports.elements.length
        ) {
            this.setState({ fetching: false });
        }
    }

    onBlockClick = (event: { target: HTMLInputElement }): void => {
        this.props.blockMessage(event.target.id);
    };

    onResolveClick = (event: { target: HTMLInputElement }): void => {
        this.props.resolveMessage(event.target.id);
    };

    renderList(): JSX.Element[] {
        return this.props.reports.elements.map(
            ({ id, state, payload: { reportType, message } }: Element) => {
                const props = {
                    id,
                    state,
                    reportType,
                    message: message === null ? '-' : message,
                    onBlockClick: this.onBlockClick,
                    onResolveClick: this.onResolveClick,
                };
                return <ListItem key={id} {...props} />;
            }
        );
    }

    render() {
        return (
            <div>
                {this.state.fetching ? 'LOADING' : null}
                <h1>Reports ({this.props.reports.size})</h1>
                {this.renderList()}
            </div>
        );
    }
}

const mapStateToProps = ({ reports }: StoreState): { reports: Report } => {
    return { reports };
};

export const ReportList = connect(mapStateToProps, {
    fetchReports,
    blockMessage,
    resolveMessage,
})(_ReportList);
