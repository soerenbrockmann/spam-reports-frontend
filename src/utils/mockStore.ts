import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const mockStore = configureStore([thunk]);

export const store = mockStore({
    reports: { size: 25, nextOffset: '', elements: [] },
});
