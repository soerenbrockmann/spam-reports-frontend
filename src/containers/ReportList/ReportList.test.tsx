import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { ReportList } from './ReportList';
import { store } from '../../utils/mockStore';

describe('<ReportList>', () => {
    it('renders without crashing', () => {
        const component = renderer.create(
            <Provider store={store}>
                <ReportList />
            </Provider>
        );
        expect(component).toMatchSnapshot();
    });
});
