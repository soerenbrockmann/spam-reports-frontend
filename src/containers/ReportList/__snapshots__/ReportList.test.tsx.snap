// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReportList> renders without crashing 1`] = `
<div>
  LOADING
  <h1>
    Reports (
    25
    )
  </h1>
</div>
`;
